[Merged by Bors] - feat(topology/algebra/uniform_ring): ring completions and algebra structures

mariainesdff opened the pull request:
If `A` is an algebra over a commutative ring `R`, so is the `uniform_space.completion` of `A`.

If `A` is a normed algebra over a normed field `𝕜`, then so is `uniform_space.completion A`.


---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

To indicate co-authors, include lines at the bottom of the commit message 
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
-->
- [x] depends on: #14846

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

mariainesdff committed:
feat(topology/algebra/uniform_ring): add algebra instance

mariainesdff committed:
lint - remove simp

ocfnash commented on src/topology/algebra/uniform_ring.lean:
```diff
@@ -150,6 +151,31 @@ instance : comm_ring (completion R) :=
       (assume a b, by rw [← coe_mul, ← coe_mul, mul_comm]),
  ..completion.ring }
 
```
How about:
```suggestion

namespace ring_completion

```
to save repeating this for each statement.

Vierkantor commented on src/topology/algebra/uniform_ring.lean:
```diff
@@ -5,6 +5,7 @@ Authors: Patrick Massot, Johannes Hölzl
 -/
 import topology.algebra.group_completion
 import topology.algebra.ring
+import algebra.algebra.basic
```
```suggestion
import algebra.algebra.basic
import topology.algebra.group_completion
import topology.algebra.ring
```

eric-wieser commented on src/topology/algebra/uniform_ring.lean:
```diff
@@ -150,6 +151,31 @@ instance : comm_ring (completion R) :=
  ..completion.ring }
 
+instance ring_completion.algebra : algebra R (completion R) :=
+(uniform_space.completion.coe_ring_hom : R →+* (completion R)).to_algebra
```
This probably produces a diamond with the nat-smul action (I think `ring_hom.to_algebra` always does)

Vierkantor replied:
Not precisely: `ℕ` and `ℤ` don't have the instances required of `R` here. The `algebra'` instance below does cause a diamond, though.

eric-wieser replied:
I think that was what I meant to comment on!

eric-wieser commented on src/topology/algebra/uniform_ring.lean:
```diff
@@ -150,6 +151,31 @@ instance : comm_ring (completion R) :=
+section algebra
+variables (S : Type*) [comm_ring S] [algebra S R]
+
+instance ring_completion.algebra' : algebra S (completion R) :=
```
You should add a lemma that says
```
(↑(algebra_map S R s) : completion R) = algebra_map S (completion R) s
```
and one that says
```
↑(s • r) = (s • ↑r : completion R)
```

mariainesdff replied:
Done!

Vierkantor commented on src/topology/algebra/uniform_ring.lean:
```diff
@@ -150,6 +151,31 @@ instance : comm_ring (completion R) :=
+
+instance ring_completion.algebra' : algebra S (completion R) :=
+((uniform_space.completion.coe_ring_hom : R →+* completion R).comp (algebra_map S R)).to_algebra
+
```
This instance has the potential to form algebra diamonds, for example:
```lean
example : (algebra_int (completion R)).smul = (ring_completion.algebra' R ℤ).smul := rfl -- fails
```
And the same for `ℕ` if we weaken the `comm_ring S` assumption above. I can't tell for sure whether you're likely to multiply integers with `completion R` (I imagine you might want to but the diamond has been lurking for longer so maybe not), but it's likely to be very painful if you do.

To fix the diamond, you should supply a custom definition of the `nsmul` and `zmul` fields for the `add_monoid`..`comm_group` instances on `completion R`, so that `n \bu x` first casts `n` to `R` before multiplying.

eric-wieser replied:
I've made an attempt at filling those fields in #14846, but using a different definition to the one you suggest.

mariainesdff replied:
Thanks for doing this! I'll try to incorporate it now.

Vierkantor commented on src/topology/algebra/uniform_ring.lean:
```diff
@@ -150,6 +151,31 @@ instance : comm_ring (completion R) :=
+rfl
+
+section algebra
+variables (S : Type*) [comm_ring S] [algebra S R]
```
We can weaken this assumption for free:
```suggestion
variables (S : Type*) [comm_semiring S] [algebra S R]
```

Vierkantor reviewed (commented):
I'm worried about the diamond, otherwise LGTM with some small tweaks.

eric-wieser committed:
chore(topology/algebra/group_completion): add lemmas about nsmul and zsmul

This also provides a general scalar action, although no lemmas hold for it generally as there is no typeclass providing a `uniform_continuous_smul` lemma.

eric-wieser committed:
Oh, we have these typeclasses after all

eric-wieser committed:
missing instances, docstring

eric-wieser committed:
tidy variables

mariainesdff committed:
suggested changes, except fix diamond

mariainesdff committed:
Merge remote-tracking branch 'origin/eric-wieser/golf-completion' into mariaines_dff/ring_completion

mariainesdff committed:
edit analysis/normed_space/completion

eric-wieser commented on src/analysis/normed_space/completion.lean:
```diff
@@ -32,7 +32,7 @@ instance : normed_space 𝕜 (completion E) :=
     (is_closed_le (continuous_const_smul _).norm (continuous_const.mul continuous_norm)) $
     λ y, by simp only [← coe_smul, norm_coe, norm_smul],
-  .. completion.module 𝕜 E }
+  ..@completion.module 𝕜 E _ _ _ _ _ (normed_space.to_has_uniform_continuous_const_smul 𝕜 E) }
```
Can you add the `normed_algebra` instance in this file too, now that it has an `algebra` instance to inherit from?

eric-wieser commented on src/topology/algebra/uniform_ring.lean:
```diff
@@ -150,8 +151,45 @@ instance : comm_ring (completion R) :=
+variables (S : Type*) [comm_semiring S] [algebra S R]
+
+instance algebra' : algebra S (completion R) :=
+((uniform_space.completion.coe_ring_hom : R →+* completion R).comp (algebra_map S R)).to_algebra
```
This should now be something like
```suggestion
{ to_ring_hom := (uniform_space.completion.coe_ring_hom : R →+* completion R).comp (algebra_map S R),
  smul_def' := sorry,
  commutes' := sorry }
```

eric-wieser committed:
Merge branch 'master' into mariaines_dff/ring_completion

eric-wieser commented on src/topology/algebra/uniform_ring.lean:
```diff
@@ -150,8 +151,45 @@ instance : comm_ring (completion R) :=
  ..completion.ring }
 
-end uniform_space.completion
+namespace ring_completion
```
This shouldn't need a new namespace, a ring completion is just a regular completion, just like a group homomorphism is just a monoid homomorphism.

eric-wieser commented on src/topology/algebra/uniform_ring.lean:
```diff
@@ -150,8 +151,45 @@ instance : comm_ring (completion R) :=
+begin
+  simp only [←mul_assoc, algebra.smul_def, map_mul],
+  congr,
+end⟩
```
I think once you make the changes above you'll find that these two declarations already exist.

mariainesdff committed:
remove namespace

mariainesdff committed:
fix diamonds

mariainesdff committed:
lint

eric-wieser commented on src/topology/algebra/uniform_ring.lean:
```diff
@@ -150,8 +156,61 @@ instance : comm_ring (completion R) :=
 
+lemma algebra_smul_eq (r : R) (x : (completion R)) :
+  r • x = (r : completion R) * x :=
+by rw algebra.smul_def; refl
```
This lemma doesn't look very useful to me; a better lemma would be that `algebra_map _ _ x = coe x`

eric-wieser commented on src/topology/algebra/uniform_ring.lean:
```diff
@@ -150,8 +156,61 @@ instance : comm_ring (completion R) :=
+  (algebra_map S R s : completion R) = algebra_map S (completion R) s :=
+rfl
+
+instance is_scalar_tower : is_scalar_tower S R (completion R) :=
```
Is this lemma true more generally without any mention of algebras as: for
```suggestion
instance is_scalar_tower [is_scalar_tower S T R] : is_scalar_tower S T (completion R) :=
```

mariainesdff committed:
generalize is_scalar_tower

eric-wieser commented on src/topology/algebra/uniform_ring.lean:
```diff
@@ -150,8 +156,66 @@ instance : comm_ring (completion R) :=
+  rw [← coe_algebra_map, ← coe_algebra_map],
+  simp only [algebra.algebra_map_eq_smul_one],
+  rw [smul_assoc, one_smul],
+end⟩
```
I suspect this is still less general than it can be. The version I'm thinking of is
```lean
instance [has_scalar N X] [has_scalar M N]
  [has_uniform_continuous_const_smul M X] [has_uniform_continuous_const_smul N X]
  [is_scalar_tower M N X] : is_scalar_tower M N (completion X) :=
⟨λ m n x, begin
  have : _ = (_ : completion X → completion X) :=
    map_comp (uniform_continuous_const_smul m) (uniform_continuous_const_smul n),
  refine eq.trans _ (congr_fun this.symm x),
  exact congr_arg (λ f, completion.map f x) (by exact funext (smul_assoc _ _)),
end⟩
```
which goes below https://github.com/leanprover-community/mathlib/blob/3dbb894f9623bb03b0f1cb2a60f9c54d7a0cfc68/src/topology/algebra/uniform_mul_action.lean#L95-L99

You should be able to write an `smul_comm_class` instance in the same way.

eric-wieser commented on src/topology/algebra/uniform_ring.lean:
```diff
@@ -31,6 +32,11 @@ open_locale classical
 noncomputable theory
 universes u
 
+lemma uniform_continuous_mul_const (α : Type*) [ring α] [uniform_space α] [uniform_add_group α]
```
```suggestion
lemma uniform_continuous_const_mul (α : Type*) [ring α] [uniform_space α] [uniform_add_group α]
```
`(*) r` is multiplication by a constant on the left, not the right.

Are you sure you can't prove this with `uniform_continuous_const_smul`?

mariainesdff committed:
add normed_algebra instance

mariainesdff committed:
changes in uniform_mul_action.lean

mariainesdff committed:
rename lemma

eric-wieser commented on src/topology/algebra/uniform_mul_action.lean:
```diff
@@ -95,9 +95,34 @@ variable [has_scalar M X]
+  exact congr_arg (λ f, completion.map f x) (by exact funext (smul_assoc _ _)),
+end⟩
+
+instance [has_scalar N X] [has_scalar M N] [smul_comm_class M N X]
```
I think the linter will tell you this in an hour:
```suggestion
instance [has_scalar N X] [smul_comm_class M N X]
```

eric-wieser commented on src/analysis/normed_space/completion.lean:
```diff
@@ -55,5 +57,43 @@ to_complₗᵢ.to_continuous_linear_map
+
+variables (𝕜) (A : Type*)
+
+instance [semi_normed_ring A] [normed_algebra 𝕜 A] : normed_ring (completion A) :=
```
Linter will complain here too:
```suggestion
instance [semi_normed_ring A] : normed_ring (completion A) :=
```

mariainesdff committed:
lint

mariainesdff committed:
add algebra_map_eq_coe

eric-wieser commented on src/topology/algebra/uniform_ring.lean:
```diff
@@ -31,6 +32,11 @@ open_locale classical
+lemma uniform_continuous_const_mul (α : Type*) [ring α] [uniform_space α] [uniform_add_group α]
+  [topological_ring α] (r : α) : uniform_continuous ((*) r) :=
+uniform_continuous_of_continuous_at_zero (smul_add_hom α α r)
+  (continuous.continuous_at (continuous_mul_left r))
```
I've extracted a variant of this to #15974

eric-wieser commented on src/topology/algebra/uniform_ring.lean:
```diff
@@ -150,8 +156,52 @@ instance : comm_ring (completion R) :=
+instance : algebra R (completion R) :=
+{ commutes' := λ r x, by simp only [ring_hom.to_fun_eq_coe, mul_comm],
+  smul_def' := λ r x, congr_fun (map_mul_eq_mul_coe R r) x,
+  .. (uniform_space.completion.coe_ring_hom : R →+* completion R) }
```
I think with #15974 in place you no longer need this instance, as it's just a special case of the one below.

eric-wieser commented on src/topology/algebra/uniform_ring.lean:
```diff
@@ -150,8 +156,52 @@ instance : comm_ring (completion R) :=
+
+lemma coe_algebra_map (s : S) :
+  (algebra_map S R s : completion R) = algebra_map S (completion R) s :=
+rfl
```
I would lean towards
```suggestion
lemma algebra_map_def (s : S) :
  algebra_map S (completion R) s = (algebra_map S R s : completion R) :=
rfl
```

eric-wieser commented on src/topology/algebra/uniform_ring.lean:
```diff
@@ -150,8 +156,52 @@ instance : comm_ring (completion R) :=
+
+@[simp] lemma algebra'_smul_eq (s : S) (x : completion R) :
+  s • x = (algebra_map S R s : completion R) * x :=
+by rw algebra.smul_def; refl
```
I would recommend just deleting this lemma; if you really need it elsewhere, it's just `rw [algebra.smul_def, completion.algebra_map_def]`.

mariainesdff committed:
suggested changes

eric-wieser committed:
Merge branch 'master' into mariaines_dff/ring_completion

eric-wieser committed:
Missed rename from `has_scalar` to `has_smul`

eric-wieser committed:
Replace more has_scalar with has_smul

mariainesdff committed:
Changes after #15974

mariainesdff committed:
missed rename

eric-wieser commented on src/topology/algebra/uniform_ring.lean:
```diff
@@ -151,8 +152,32 @@ instance : comm_ring (completion R) :=
+lemma algebra_map_def (s : S) :
+  algebra_map S (completion R) s = (algebra_map S R s : completion R) :=
+rfl
+
```
It's probably worth adding
```
/-- A shortcut instance for the common case -/
instance algebra' : algebra R (completion R) :=
by apply_instance
```
mainly because this confirms that the instance above really does generalize how we expect it to.

eric-wieser commented on src/topology/algebra/uniform_ring.lean:
```diff
@@ -151,8 +152,32 @@ instance : comm_ring (completion R) :=
+    apply_instance, },
+end
 
+instance : algebra S (completion R) :=
```
Doesn't this work when `R` is non-commutative too?

eric-wieser commented on src/topology/algebra/uniform_ring.lean:
```diff
@@ -151,8 +152,32 @@ instance : comm_ring (completion R) :=
+end
 
+instance : algebra S (completion R) :=
+{ commutes' := λ s x, by simp only [ring_hom.to_fun_eq_coe, mul_comm],
```
This proof will need to change to use `algebra.commutes` instead of `mul_comm`.

eric-wieser replied:
That looks like
```suggestion
{ commutes' := λ s x, completion.induction_on x
    (is_closed_eq (continuous_mul_left _) (continuous_mul_right _)) $ λ r,
      by simpa only [coe_mul] using congr_arg (coe : R → completion R) (algebra.commutes s r),
```

mariainesdff committed:
suggested changes

eric-wieser commented on src/topology/algebra/uniform_ring.lean:
```diff
@@ -151,8 +154,36 @@ instance : comm_ring (completion R) :=
+  refine completion.induction_on x _ (λ r, _),
+  { exact is_closed_eq (completion.continuous_map) (continuous_mul_left _) },
+  { rw [map_coe (has_uniform_continuous_const_smul.uniform_continuous_const_smul s),
+      algebra.smul_def, coe_mul];
```
```suggestion
  { rw [map_coe (uniform_continuous_const_smul s) r, algebra.smul_def, coe_mul] },
```

mariainesdff committed:
noncommutativity

eric-wieser commented on src/topology/algebra/uniform_ring.lean:
```diff
@@ -142,6 +145,33 @@ instance top_ring_compl : topological_ring (completion α) :=
 
+section algebra
+variables (R : Type*) [ring R] [uniform_space R] [uniform_add_group R] [topological_ring R]
+  (S : Type*) [comm_semiring S] [algebra S R] [has_uniform_continuous_const_smul S R]
```
Can you rename the variables in this section from `S R` to `R A`, since that matches what we normally do for `algebra`, and also matches what you wrote in the PR description?

eric-wieser commented on src/topology/algebra/uniform_ring.lean:
```diff
@@ -25,6 +26,8 @@ the main constructions deal with continuous ring morphisms.
   to a complete separated group `S` to `completion R`.
 * `uniform_space.completion.map_ring_hom` : promotes a continuous ring morphism
   from `R` to `S` into a continuous ring morphism from `completion R` to `completion S`.
+
```
Can you update the docstring on lines 13-16 to mention the `algebra` structure too?

eric-wieser reviewed (approved):
Thanks for adding the non-commutative version. I'm very happy with this now :)

mariainesdff committed:
rename + docstring
